core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -666,24 +667,22 @@ static class JoinConditionBasedPredicateInference {
       equivalence = BitSets.closure(equivalence);
     }
 
+
```

Comment:
While I was debugging, I refactored `inferPredicates` into several helper functions to help me understand what was going on. I think that these changes are good, but it does complicates the git diff. I've left comments to highlight the real changes.

If needed, I can remove the refactoring from this PR and do it separately.